core/src/main/java/org/apache/iceberg/MicroBatches.java

```diff
@@ -0,0 +1,266 @@
+                                          int startFileIndex, long targetSizeInBytes, boolean isStarting) {
+      if (indexedManifests.isEmpty()) {
+        return new MicroBatch(snapshot.snapshotId(), startFileIndex, startFileIndex + 1, 0L,
+            Collections.emptyList(), true);
+      }
+
+      long currentSizeInBytes = 0L;
+      int currentFileIndex = 0;
+      boolean isLastIndex = false;
+      List<FileScanTask> tasks = Lists.newArrayList();
+
+      for (int idx = 0; idx < indexedManifests.size(); idx++) {
+        currentFileIndex = indexedManifests.get(idx).second();
+
+        try (CloseableIterable<FileScanTask> taskIterable = open(indexedManifests.get(idx).first(), isStarting);
+            CloseableIterator<FileScanTask> taskIter = taskIterable.iterator()) {
+
+          while (taskIter.hasNext()) {
+            FileScanTask task = taskIter.next();
+            if (currentFileIndex >= startFileIndex) {
+              // Make sure there's at least one task in each MicroBatch to void job to be stuck, always add task
+              // firstly.
+              tasks.add(task);
+              currentSizeInBytes += task.length();
+            }
+
+            currentFileIndex++;
+            if (currentSizeInBytes >= targetSizeInBytes) {
+              break;
+            }
+          }
+
+          if (idx + 1 == indexedManifests.size() && !taskIter.hasNext()) {
+            // If this is the last file scan task in last manifest, set the flag to true.
+            isLastIndex = true;
+          }
+        } catch (IOException ioe) {
+          LOG.warn("Failed to close task iterable", ioe);
+        }
+
+        if (currentSizeInBytes >= targetSizeInBytes) {
+          if (tasks.size() > 1 && currentSizeInBytes > targetSizeInBytes) {
+            // If there's more than 1 task in this batch, and the size exceeds the limit, we should revert last
+            // task to make sure we don't exceed the size limit.
+            FileScanTask extraTask = tasks.remove(tasks.size() - 1);
+            currentSizeInBytes -= extraTask.length();
+            currentFileIndex--;
+            isLastIndex = false;
+          }
+
+          break;
+        }
+      }
+
+      return new MicroBatch(snapshot.snapshotId(), startFileIndex, currentFileIndex, currentSizeInBytes,
+          tasks, isLastIndex);
+    }
+
+    private CloseableIterable<FileScanTask> open(ManifestFile manifestFile, boolean isStarting) {
+      CloseableIterable<FileScanTask> fileScanTasks = null;
+      if (isStarting) {
+        fileScanTasks = new ManifestGroup(io, ImmutableList.of(manifestFile))
+            .specsById(specsById)
+            .caseSensitive(caseSensitive)
+            .planFiles();
+
+      } else {
+        Iterable<ManifestFile> newManifests = Iterables.filter(ImmutableList.of(manifestFile),
+            manifest -> manifest.snapshotId() == snapshot.snapshotId());
```

Comment:
This is suspicious because it creates a list of a single element, then filters. I think this should be replaced with an `if` that either returns the `ManifestGroup` result, or `CloseableIterable.empty()` if the snapshot ID doesn't match.

Reply:
Also, I think the snapshot ID is already guaranteed to match because manifests are filtered in `generate` when `isStarting` is false.

Reply:
OK, I will fix it.